specs/CDPSessionIdSupport.md

```diff
@@ -0,0 +1,328 @@
+
+However, dedicated web workers are not returned from `'Target.getTargets'`, and you have to call DevToolsProtocol method `Target.setAutoAttach` to be able to attach to them.
+
+And shared worker is separate from any page or iframe target, and therefore will not be auto attached. You have to call `Target.attachToTarget` to attach to them. The shared workers can be enumerated with `Target.getTargets`. They are also discoverable, that is, you can call `Target.setDiscoverTargets` to receive `Target.targetCreated` event when a shared worker is created.
+
+To summarize, there are two ways of finding the targets and neither covers all 3 scenarios.
+
+|      | Web pages | Dedicated web workers | Shared workers |
+| --- | --- | --- | --- |
+| Target.getTargets, Target.setDiscoverTargets, Target.targetCreated, Target.attachToTarget | ✔ | | ✔ |
+| Target.setAutoAttach, Target.attachedToTarget | ✔ | ✔ |  |
+
+# Examples
+
+The example below illustrates how to collect messages logged by console.log calls by JavaScipt code from various parts of the web page, including dedicated web worker.
+
+## Win32 C++
+```cpp
+
+void MyApp::HandleDevToolsProtocalPTargets()
+{
+    // The sessions and targets descriptions are tracked by 2 maps:
+    // SessionId to TargetId map:
+    //   std::map<std::wstring, std::wstring> m_devToolsSessionMap;
+    // TargetId to description map, where description is "<target type>,<target url>".
+    //   std::map<std::wstring, std::wstring> m_devToolsTargetInfoMap;
+    // GetJSONStringField is a helper function that can retrieve a string field from a json message.
+  
+    wil::com_ptr<ICoreWebView2DevToolsProtocolEventReceiver> receiver;
+    // Listen to Runtime.consoleAPICalled event which is triggered when console.log is called by script code.
+    CHECK_FAILURE(
+        m_webView->GetDevToolsProtocolEventReceiver(L"Runtime.consoleAPICalled", &receiver));
+    CHECK_FAILURE(receiver->add_DevToolsProtocolEventReceived(
+        Callback<ICoreWebView2DevToolsProtocolEventReceivedEventHandler>(
+            [this](
+                ICoreWebView2* sender,
+                ICoreWebView2DevToolsProtocolEventReceivedEventArgs* args) -> HRESULT
+            {
+                // Get console.log message details and which target it comes from
+                wil::unique_cotaskmem_string parameterObjectAsJson;
+                CHECK_FAILURE(args->get_ParameterObjectAsJson(&parameterObjectAsJson));
+                std::wstring eventSource;
+                std::wstring eventDetails = parameterObjectAsJson.get();
+                wil::com_ptr<ICoreWebView2DevToolsProtocolEventReceivedEventArgs2> args2;
+                if (SUCCEEDED(args->QueryInterface(IID_PPV_ARGS(&args2))))
+                {
+                    wil::unique_cotaskmem_string sessionId;
+                    CHECK_FAILURE(args2->get_SessionId(&sessionId));
+                    if (sessionId.get() && *sessionId.get())
```

Comment:
Do we need to check for null pointer coming out of `get_SessionId`? I thought the general rule was to use empty strings, not null pointers.

Reply:
We don't have to. The implementation always return empty string

Reply:
Update if condition above to just check for empty string rather than nullptr.